include/ginkgo/core/matrix/dense.hpp

```diff
@@ -401,6 +405,191 @@ class Dense
      */
     void fill(const ValueType value);
 
+    /**
+     * Creates a permuted copy $A'$ of this matrix $A$ with the given
+     * permutation $P$. By default, this computes a symmetric permutation
+     * (permute_mode::symmetric). For the effect of the different permutation
+     * modes, see the following table.
+     *
+     * mode              | entry mapping              | matrix representation
+     * ------------------|----------------------------|----------------------
+     * none              | $A'(i, j) = A(i, j)$       | $A' = A$
+     * rows              | $A'(i, j) = A(p[i], j)$    | $A' = P A$
+     * columns           | $A'(i, j) = A(i, p[j])$    | $A' = A P^T$
+     * inverse_rows      | $A'(p[i], j) = A(i, j)$    | $A' = P^{-1} A$
+     * inverse_columns   | $A'(p[i], j) = A(i, j)$    | $A' = A P^{-T}$
+     * symmetric         | $A'(i, j) = A(p[i], p[j])$ | $A' = P A P^T$
+     * inverse_symmetric | $A'(p[i], p[j]) = A(i, j)$ | $A' = P^{-1} A P^{-T}$
+     *
+     * @param permutation  The input permutation.
+     * @param mode  The permutation mode. If permute_mode::inverse is set, we
+     *              use the inverse permutation $P^{-1}$ instead of $P$.
+     *              If permute_mode::rows is set, the rows will be permuted.
+     *              If permute_mode::columns is set, the columns will be
+     *              permuted.
+     * @return  The permuted matrix.
+     */
+    std::unique_ptr<Dense> permute(
```

Comment:
templates and `ptr_param` don't go together :(

Reply:
:(((

Reply:
Yes, unfortunately templates need exact type matches, otherwise the search space would be just too great, if you allowed implicit conversions or even chains thereof.